### PR TITLE
feat: call `aggsender validator` in regular `aggsender`

### LINKED ISSUE #730: aggsender: PP allow to generate a second signature using `aggsender-validator`
# Feature Request
Aggsender can configure that require a extra validation

### PR DESCRIPTION
**Pull request by goran-ethernal**
## Description

This PR implements the real call to `Aggsender Validator` in the regular `Aggsender`.

It also adds the validator signature to the `submitCertificate` request to the `Agglayer` by putting it into the request headers under key: `x-agglayer-extra-certificate-signature`.

### ⚙️ Config Changes
- Added `RequireValidatorCall` to the `Aggsender` config which defines if a call to the `Aggsender Validator` is required before sending the certificate to the `Agglayer`.
- Added `ValidatorClient` config to setup `grpc` client to the `Aggsender Validator` which is in the form of all the `grpc` clients config:
```
[AggSender]
RequireValidatorCall = false
[AggSender.ValidatorClient]
URL = "validatorURL"
MinConnectTimeout = "5s"
RequestTimeout = "300s"
UseTLS = false
[AggSender.ValidatorClient.Retry]
	InitialBackoff = "1s"
	MaxBackoff = "10s"
	BackoffMultiplier = 2.0
	MaxAttempts = 20
```

Fixes #730

### CODE REVIEW

File: agglayer/grpc/agglayer_grpc_client.go [hunk 1]
```diff
@@ -72,6 +77,13 @@ func (a *AgglayerGRPCClient) SendCertificate(ctx context.Context,
 	ctx, cancel := context.WithTimeout(ctx, a.cfg.RequestTimeout.Duration)
 	defer cancel()
 
+	if validatorSignature != nil {
```
Comment (Stefan-Ethernal) [inline agglayer/grpc/agglayer_grpc_client.go]: Maybe it slightly more idiomatic, because it is a slice?

```suggestion
	if len(validatorSignature) > 0 {
```

**Review by Stefan-Ethernal — APPROVED** [review]

File: agglayer/grpc/agglayer_grpc_client.go (hunk 1 above)
Comment (goran-ethernal) [reply agglayer/grpc/agglayer_grpc_client.go]: Fix: https://github.com/agglayer/aggkit/pull/738/commits/4de1772703bb2b9a6795ef3ab4e61c4f22f8dd1d

File: aggsender/validator/validator_client.go [hunk 2]
```diff
@@ -8,16 +8,19 @@ import (
...
 
+var _ types.ValidatorClient = (*ValidatorClient)(nil)
+
 // ValidatorClient encapsulates the gRPC client and configuration
 // required to interact with the AggsenderValidator service.
 type ValidatorClient struct {
-	client        v1.AggsenderValidatorClient
-	grpcClientCfg *grpc.ClientConfig
+	client v1.AggsenderValidatorClient
+	cfg    *grpc.ClientConfig
```
[the remark this answers is not in the document]
Comment (goran-ethernal) [reply aggsender/validator/validator_client.go]: Fixed: https://github.com/agglayer/aggkit/pull/738/commits/c4481680266f26d6d2122908899893a55b065935

File: agglayer/grpc/agglayer_grpc_client.go [hunk 3]
```diff
@@ -72,6 +77,13 @@ func (a *AgglayerGRPCClient) SendCertificate(ctx context.Context,
 	ctx, cancel := context.WithTimeout(ctx, a.cfg.RequestTimeout.Duration)
 	defer cancel()
 
+	if len(validatorSignature) > 0 {
+		// Add validator signature to metadata
+		sigHex := hex.EncodeToString(validatorSignature)
+		md := metadata.Pairs(validatorSignatureMetadataKey, sigHex)
+		ctx = metadata.NewOutgoingContext(ctx, md)
```
[the remark this answers is not in the document]
Comment (goran-ethernal) [reply agglayer/grpc/agglayer_grpc_client.go]: Fixed: https://github.com/agglayer/aggkit/pull/738/commits/c4481680266f26d6d2122908899893a55b065935

File: aggsender/config/config.go [hunk 4]
```diff
@@ -104,3 +111,22 @@ func (c Config) String() string {
 		"SovereignRollupAddr: " + c.SovereignRollupAddr.Hex() + "\n" +
 		"RequireNoFEPBlockGap: " + fmt.Sprintf("%t", c.RequireNoFEPBlockGap) + "\n"
 }
+
+// Validate checks if the configuration is valid
+func (c Config) Validate() error {
+	if c.RequireValidatorCall && c.ValidatorClient.URL == "" {
+		return errValidatorClientURLNotSet
```
[the remark this answers is not in the document]
Comment (goran-ethernal) [reply aggsender/config/config.go]: Fixed: https://github.com/agglayer/aggkit/pull/738/commits/c4481680266f26d6d2122908899893a55b065935

File: aggsender/aggsender.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ type AggSender struct {
 
 	l2OriginNetwork uint32
 
-	validator types.CertificateValidateAndSigner
```
Comment (joanestebanr) [inline aggsender/aggsender.go]: Why change this type? the idea of the validator was to encapsulate all the behaviour of the extra signature. The code is already done in aggsender you just need to implement a class as [LocalValidator](https://github.com/agglayer/aggkit/blob/f824abd25a8d34091c2284a71fff75a06df93759/aggsender/validator/local_validator.go#L17) that have all the new logic

The idea is to keep as separated as possible, and also allow to choose the local one (to keep testing the code), the remote one or none

**Review by vcastellm — COMMENTED** [review]
Not sure if I understand it correctly, the proposer may need to call different validators not just one, and with this config it looks like it's only prepared to call one validator, am I missing something?

File: aggsender/aggsender.go (hunk 5 above)
Comment (goran-ethernal) [reply aggsender/aggsender.go]: Fix: https://github.com/agglayer/aggkit/pull/738/commits/9b690354e4a02a24f61ce9dbb68ef8e9a6f72b1d

**Comment by goran-ethernal** [comment]
> Not sure if I understand it correctly, the proposer may need to call different validators not just one, and with this config it looks like it's only prepared to call one validator, am I missing something?

@vcastellm For now, we go with only one validator for the TAC network. This is not an implementation of the multisig. Multisig will be done in a different manner.

**Review by vcastellm — APPROVED** [review]
LGTM

File: config/default.go [hunk 6]
```diff
@@ -191,6 +191,7 @@ RequireOneBridgeInPPCertificate = false
 RollupManagerAddr = "{{L1Config.polygonRollupManagerAddress}}"
 RollupCreationBlockL1 = {{rollupCreationBlockNumber}}
 MaxL2BlockNumber = 0
+RequireValidatorCall = false
```
Comment (joanestebanr) [inline config/default.go]: do you mind to set this below `StopOnFinishedSendingAllCertificates`? the point is that `MaxL2BlockNumber` and `StopOnFinishedSendingAllCertificates` aare related

File: cmd/run.go [hunk 7]
```diff
@@ -293,19 +294,28 @@ func createAggSender(
 	if err != nil {
 		return nil, fmt.Errorf("failed to create AggSender: %w", err)
 	}
-	// TODO: Remove this, just for testing
-	if cfg.Mode == "PessimisticProof" {
-		validator, err := createAggSenderValidator(ctx, cfg, l1InfoTreeSync,
-			l2Syncer, l1EthClient, l2Client, rollupDataQuerier)
+
+	var validator aggsendertypes.CertificateValidateAndSigner
+	if cfg.RequireValidatorCall {
```
Comment (joanestebanr) [inline cmd/run.go]: The `validator` only supports `PessimisticProof`. So you can't set LocalValidator if it's another mode, and make sense that if `cfg.RequireValidatorCall` is `true` check that mode is the expected one

File: config/default.go (hunk 6 above)
Comment (goran-ethernal) [reply config/default.go]: Fix: https://github.com/agglayer/aggkit/pull/738/commits/c3f5c90b0ef721fa1d5eada554ac726dae9ab001

File: cmd/run.go (hunk 7 above)
Comment (goran-ethernal) [reply cmd/run.go]: Fix: https://github.com/agglayer/aggkit/pull/738/commits/c3f5c90b0ef721fa1d5eada554ac726dae9ab001

**Review by joanestebanr — APPROVED** [review]
